### PR TITLE
Perform naiive branch relaxing (and other fixes to the BIR passes)

### PR DESCRIPTION
**Pull request by bmourad01**
There's a number of changes here:

1. Branch relaxing (see #195)
2. Hacky workaround for #179 (only affects Thumb, see the huge comment in `Bir_passes.Shape.split_on_conditional`)
3. Only "finalize" higher variables at the "implicit exit" node. If we have a block that does `goto L_0x1234`, then this is not an implicit exit and we can assume that the higher vars aren't needed "at exit" in any particular register.
4. Adds the `lsl rt, [rn, rm, lsl #n]` pattern to the current selector (results in much smaller code!)

### CODE REVIEW

File: bap-vibes/src/bir_passes.mli [hunk 1]
```diff
@@ -57,4 +57,4 @@ end
 (** [run patch] creates the BIR from [patch], then applies a series of
     transformations to it. The resulting code is then ready to be handed
     off to the instruction selector. *)
-val run : Data.Patch.t -> t KB.t
+val run : Data.Patch.t -> patch_spaces:Data.Patch_space_set.t -> t KB.t
```
[the remark this answers is not in the document]
Comment (bmourad01) [reply bap-vibes/src/bir_passes.mli:60]: I agree, though there are limitations to doing it later down the pipeline. If we did this at the level of the patcher, where the code has already been generated, then we would need to intercept the error from the assembler and then manually adjust the generated assembly code (or the VIBES IR). Doing it at the BIR level is easier.

**Comment by philzook58** [comment]
This also adds the "init-mem" feature?

File: bap-vibes/src/config.ml [hunk 2]
```diff
@@ -184,6 +184,7 @@ let wp_params_to_string (wp_params : Wp_params.t) : string =
         "user-fun-specs-orig: %s" (triple_list wp_params.user_func_specs_orig);
       Printf.sprintf
         "user-fun-specs-mod: %s" (triple_list wp_params.user_func_specs_mod);
+      Printf.sprintf "init-mem: %b" wp_params.init_mem;
```
Comment (philzook58) [inline bap-vibes/src/config.ml:187]: Does this PR enable the "init-mem" feature?

**Comment by philzook58** [comment]
Seems like good stuff. Great work! I am worried that it feels like the patch_spaces being in bir_passes is ensnarling the codebase in itself, but I don't see an easy way around it. Everything _is_ dependent on everything if you want it to be really good unfortunately.

Is point 3 above correct? I think this is maybe because of a weakness of the schema of the config file. Each possible exit point should have it's own notion in general of where hvars need to be put. In other words we need a write-to relation:  (addr, hvar, lowloc)

**Comment by bmourad01** [comment]
Yes, it seems that this would be an avenue for future work. In general when we have a patch like:

```
x = y + z;
if (x > w) {
  goto L_0x1234;
}
--y;
```

We could just assume that once we go to `0x1234` there is an entirely different set of data dependencies (but for our purposes we just assume there are none at all), since we're kind of "aborting" the normal control flow of the function being patched. 

Meanwhile, the fallthrough to `--y` means that we're continuing back to the end of the patch point, which means the dependencies should be preserved.

It would be nice to have some granularity per-exit about what kind of data dependency is needed.

**Pull request merged**